### PR TITLE
PARQUET-1663: [C++] Provide API to check the presence of repeated fields

### PR DESCRIPTION
**Pull request by czxrrr**
added two functions
hasMapType() and hasArrayType()
in case some other tools that does not support array or map can know if map or array exist in this parquet file.

### CODE REVIEW

File: cpp/src/parquet/schema.h [hunk 1]
```diff
@@ -434,6 +434,9 @@ class PARQUET_EXPORT SchemaDescriptor {
   /// PrimitiveNode. Returns -1 if not found
   int GetColumnIndex(const schema::PrimitiveNode& node) const;
 
+  bool hasArrayType() const { return has_array_type_; }
```
Comment (emkornfield) [inline cpp/src/parquet/schema.h]: methods should either be HasArrayType or has_array_type

File: cpp/src/parquet/schema_test.cc [hunk 2]
```diff
@@ -758,6 +758,50 @@ TEST_F(TestSchemaDescriptor, BuildTree) {
...
+                               ConvertedType::LIST));
+  NodePtr bag(GroupNode::Make("bag", Repetition::OPTIONAL, {list}));
+  fields.push_back(bag);
+
+  schema = GroupNode::Make("schema", Repetition::REPEATED, fields);
+  descr_.Init(schema);
+  ASSERT_EQ(true, descr_.hasArrayType());
+  ASSERT_EQ(false, descr_.hasMapType());
+
+  // 3-level list encoding
```
Comment (emkornfield) [inline cpp/src/parquet/schema_test.cc]: is this comment accurate?

Comment (emkornfield) [inline cpp/src/parquet/schema_test.cc]: does fields need to be shared across test cases?

**Comment by wesm** [comment]
Would it be sufficient to instead have a method indicating if there are any repeated child fields of the schema at all?

File: cpp/src/parquet/schema_test.cc (hunk 2 above)
Comment (czxrrr) [reply cpp/src/parquet/schema_test.cc]: There are two test cases, they are different, so not sharing.

**Comment by emkornfield** [comment]
@czxrrr can you respond to Wes's comment?

**Comment by wesm** [comment]
@czxrrr following up about my question

**Comment by wesm** [comment]
I rebased and changed to have a single method detecting the presence of repeated fields.

@emkornfield if you could give this a brief look that would be helpful

@czxrrr is this sufficient for your needs or do you need to distinguish between the presence of arrays and presence of maps?

File: cpp/src/parquet/schema_test.cc [hunk 3]
```diff
@@ -799,6 +799,48 @@ TEST_F(TestSchemaDescriptor, BuildTree) {
...
+  descr_.Init(schema);
+  ASSERT_EQ(true, descr_.HasRepeatedFields());
+  ASSERT_EQ(false, descr_.HasRepeatedFields());
+
+  // 3-level list encoding
+  NodePtr item_key = Int64("key", Repetition::REQUIRED);
+  NodePtr item_value = Boolean("value", Repetition::OPTIONAL);
+  NodePtr map(GroupNode::Make("map", Repetition::REPEATED, {item_key, item_value},
+                              ConvertedType::MAP));
+  NodePtr myMap(GroupNode::Make("myMap", Repetition::OPTIONAL, {map}));
```
Comment (emkornfield) [inline cpp/src/parquet/schema_test.cc]: nit my_map

File: cpp/src/parquet/schema_test.cc [hunk 4]
```diff
@@ -799,6 +799,48 @@ TEST_F(TestSchemaDescriptor, BuildTree) {
   ASSERT_EQ(nleaves, descr_.num_columns());
 }
 
+TEST_F(TestSchemaDescriptor, ComplexTypeCheck) {
```
Comment (emkornfield) [inline cpp/src/parquet/schema_test.cc]: might want to rename this to HasRepeatedFieldCheck

**Comment by emkornfield** [comment]
Overall looks OK to me, a couple of small comments.

**Comment by emkornfield** [comment]
LGTM, I'll wait till monday to see if this seems good @czxrrr and then merge if there is no response.

**Comment by wesm** [comment]
+1

**Comment by czxrrr** [comment]
Sorry for the late reply. Thanks!

**Pull request closed without merging**